### PR TITLE
improve ABC sampler

### PR DESCRIPTION
**Pull request by aloctavodia**
Simplify the specification of the abc sampler as previously we had a distance and a kernel with was redundant. Now we only have a gauss_kernel as the only one distance (in a future PR I will probably add more distances, but I need to do some reading and testing before that). This also adds more flexibility to define summary statistics, including user defined statistics. Finally this allows to define more complex models as the simulator can be feed with a subset of the prior parameters.

### CODE REVIEW

**Review by AlexAndorra — CHANGES_REQUESTED** [review]
I'm no expert on the sampler _per se_, so I'll defer to more experienced people on the guts of the sampler.
With that in mind, this looks good to me and will be useful when porting chapter 16 of _Rethinking 2_ 😉 
I just added some comments on the NB, with the goal of gradually standardizing our NBs galleries.

File: docs/source/notebooks/SMC-ABC_Lotka-Volterra_example.ipynb
Comment (AlexAndorra) [inline docs/source/notebooks/SMC-ABC_Lotka-Volterra_example.ipynb]: <p>With the goal of gradually standardizing the NBs format, maybe you can use the import statements convention we use on the resources repo? And also use ArviZ style?</p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3940/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (AlexAndorra) [inline docs/source/notebooks/SMC-ABC_Lotka-Volterra_example.ipynb]: <p>Maybe add a last cell with <code>watermark</code> output, like in resources repo? Finally, can you run <code>black_nbconvert</code> please?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/pymc-devs/pymc3/pull/3940/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='24'/>

File: pymc3/smc/sample_smc.py
Comment (michaelosthege) [inline pymc3/smc/sample_smc.py:77]: These breaking changes should be mentioned in the release notes

File: pymc3/distributions/simulator.py
Comment (michaelosthege) [inline pymc3/distributions/simulator.py]: Doesn't this break existing implementations in a non-obvious way?

Maybe instead do `(self, function, **args, params=None, **kwargs)` ?

File: pymc3/smc/smc.py
Comment (michaelosthege) [inline pymc3/smc/smc.py:419]: Same comment as above, regarding compatibility…
You can also use `(…, function, *, params, ...)` to require everything behind `,* , ` as a kwarg.

**Review by michaelosthege — CHANGES_REQUESTED** [review]
I'm not familiar with the sampler either, but I assume you've tested it thoroughly?

The breaking API changes should be mentioned in the release notes.

**Comment by aloctavodia** [comment]
> I'm no expert on the sampler _per se_, so I'll defer to more experienced people on the guts of the sampler.
> With that in mind, this looks good to me and will be useful when porting chapter 16 of _Rethinking 2_ wink
> I just added some comments on the NB, with the goal of gradually standardizing our NBs galleries.

It should work, but I guess pm.ode will work better.

File: pymc3/smc/smc.py
Comment (aloctavodia) [reply pymc3/smc/smc.py:419]: I will change it. I doubt the ABC sampler is widely used. But I have plans to change that in the short-term :-)

**Comment by aloctavodia** [comment]
Thanks for the comments!

**Review by AlexAndorra — APPROVED** [review]
All good now, thanks @aloctavodia 👌 
I'll let @michaelosthege merge once he approves the changes

**Review by michaelosthege — APPROVED** [review]

**Pull request merged**
